### PR TITLE
Add OAuth2SSOAuthenticator + empty plugin

### PR DESCRIPTION
**Pull request by DaazKu**
Closes https://github.com/vanilla/vanilla/issues/6602

- Add OAuth2SSOAuthenticator which allows OAuth2 SSO connexions to work with /authenticate

How to test:

Here are the providers I used to test. Because of the redirects verification you have to have your main vanilla instance run on `dev.vanilla.localhost`.
```sql
INSERT INTO `GDN_UserAuthenticationProvider` (`AuthenticationKey`, `AuthenticationSchemeAlias`, `Name`, `URL`, `AssociationSecret`, `AssociationHashMethod`, `AuthenticateUrl`, `RegisterUrl`, `SignInUrl`, `SignOutUrl`, `PasswordUrl`, `ProfileUrl`, `Attributes`, `Active`, `IsDefault`)
VALUES
	('oauth2', 'oauth2', 'oauth2', NULL, '1234', NULL, NULL, '', 'https://sso.vanilla.localhost', '', NULL, 'https://sso.vanilla.localhost/oauth2/profile', '{\"AssociationKey\":\"stub_oauth2\",\"AuthorizeUrl\":\"https://sso.vanilla.localhost/oauth2/authorize\",\"TokenUrl\":\"https://sso.vanilla.localhost/oauth2/access_token?XDEBUG_SESSION=PHPSTORM\",\"AcceptedScope\":\"profile\",\"ProfileKeyEmail\":\"Email\",\"ProfileKeyPhoto\":\"\",\"ProfileKeyName\":\"Name\",\"ProfileKeyFullName\":\"Name\",\"ProfileKeyUniqueID\":\"UserID\",\"BearerToken\":false,\"BaseUrl\":\"https://sso.vanilla.localhost\"}', 1, 0),
	('oauth2-bak', 'oauth2-bak', NULL, NULL, '8ZosRm5QeBCO06hKH4io3TsfO7lxVppH01HButehUweSi3fzSpFOPGzPKyaKDqCe', NULL, NULL, '', 'https://vanilla.auth0.com', '', NULL, 'https://vanilla.auth0.com/userinfo', '{\"AssociationKey\":\"C61DLFgy5lik_7X_mPVpRdphK1FxeZYI\",\"AuthorizeUrl\":\"https://vanilla.auth0.com/authorize?audience=https%3A%2F%2Fvanilla.auth0.com%2Fapi%2Fv2%2F\",\"TokenUrl\":\"https://vanilla.auth0.com/oauth/token\",\"AcceptedScope\":\"openid profile email\",\"ProfileKeyEmail\":\"email\",\"ProfileKeyPhoto\":\"picture\",\"ProfileKeyName\":\"name\",\"ProfileKeyFullName\":\"\",\"ProfileKeyUniqueID\":\"sub\",\"BearerToken\":false,\"BaseUrl\":\"https://vanilla.auth0.com\"}', 1, 0);
```
OAuth2 does not allow multiple providers to be created right now. You'll have to switch AuthenticationKey and AuthenticationSchemeAlias to test the other provider.

One provider is Auth0
The other is in https://github.com/vanilla/stub-sso-providers

**Note that the OAuth2 icon is not visible.** 

**The plugin does not need to be reviewed.**

### CODE REVIEW

File: plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
+ */
+
+class OAuth2AuthenticatorPlugin extends Gdn_Plugin {
+
+    /**
+     * Inject sign-in button into the sign in page.
+     *
+     * @param EntryController $sender.
+     * @param EntryController $args.
```
Comment (initvector) [inline plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php]: This is usually an array.

File: plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     *
+     * @return mixed|bool Return null if not configured
```
Comment (initvector) [inline plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php]: It doesn't return anything.

File: plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            // Add the sign in button method to the controller.
+            $method = [
+                'Name' => 'oauth2',
+                'SignInHtml' => $this->signInButton()
+            ];
+
+            $sender->Data['Methods'][] = $method;
+        }
+    }
+    /**
```
Comment (initvector) [inline plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php]: Need some space here.

File: plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+     * Inject a sign-in icon into the ME menu.
+     *
+     * @param Gdn_Controller $sender.
+     * @param Gdn_Controller $args.
```
Comment (initvector) [inline plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php]: `$args` is usually an array.

File: plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public function base_beforeSignInButton_handler($sender, $args) {
+        echo ' '.$this->signInButton('icon').' ';
+    }
+
+
```
Comment (initvector) [inline plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php]: Need some kinda doc block here.

File: plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        $redirect_uri = '/authenticate/oauth2';
+        $defaultParams = [
+            'response_type' => 'code',
+            'client_id' => val('AssociationKey', $provider),
+            'redirect_uri' => url($redirect_uri, true),
+            'scope' => val('AcceptedScope', $provider)
+        ];
+
+        $url =  $uri.'?'.http_build_query($defaultParams);
+
```
Comment (initvector) [inline plugins/oauth2authenticator/OAuth2AuthenticatorPlugin.php]: One too many empty lines.

File: plugins/oauth2authenticator/src/Vanilla/Authenticator/OAuth2Authenticator.php [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+        $this->providerData = $authProviderModel->getProviderByKey($authenticatorID);
+
+        if (empty($this->providerData)) {
+            throw new Exception('Empty providerData in OAuth2Authenticator.');
+        }
+    }
+
+    /**
+     * @inheritDoc
```
Comment (initvector) [inline plugins/oauth2authenticator/src/Vanilla/Authenticator/OAuth2Authenticator.php]: I believe the `@inheritDoc` tag is required to be wrapped in curly braces: `{@inheritDoc}`

**Comment by DaazKu** [comment]
Deleted a bunch of review on the plugin which should not be reviewed at this moment.

**Pull request merged**